### ISSUE TITLE
Issues with extensible matching in AD

### ISSUE THREAD

**Issue by ve7cxz** [cannatag/ldap3#447]
Hi,

I'm trying to use extensible matching to get a list of users who are members of an AD group - in this case, I'm trying to find anyone who may be a member of the Domain Admins group either directly or via membership of another group.

My search looks like this:-

        conn.search(
            search_base=BASE_DN,
            search_filter='(&(objectClass=user)(memberOf:1.2.840.113556.1.4.1941:=CN=Domain Admins,CN=Users,DC=example,DC=com'),
            search_scope=SUBTREE,
            attributes = ALL_ATTRIBUTES
        )

When run, I get this stack trace:-

```
Traceback (most recent call last):
<...snip outer code...>
  File "/path/to/homedir/venv/lib/python3.4/site-packages/ldap3/core/connection.py", line 762, in search
    self.server.schema if self.server else None)
  File "/path/to/homedir/venv/lib/python3.4/site-packages/ldap3/operation/search.py", line 370, in search_operation
    request['filter'] = compile_filter(parse_filter(search_filter, schema, auto_escape, auto_encode).elements[0])  # parse the searchFilter string and compile it starting from the root node
  File "/path/to/homedir/venv/lib/python3.4/site-packages/ldap3/operation/search.py", line 230, in compile_filter
    boolean_filter[pos] = compile_filter(element)
  File "/path/to/homedir/venv/lib/python3.4/site-packages/ldap3/operation/search.py", line 267, in compile_filter
    matching_filter['dnAttributes'] = DnAttributes(filter_node.assertion['dnAttributes'])
  File "/path/to/homedir/venv/lib/python3.4/site-packages/pyasn1/type/univ.py", line 103, in __init__
    base.AbstractSimpleAsn1Item.__init__(self, value, **kwargs)
  File "/path/to/homedir/venv/lib/python3.4/site-packages/pyasn1/type/base.py", line 240, in __init__
    value = self.prettyIn(value)
  File "/path/to/homedir/venv/lib/python3.4/site-packages/pyasn1/type/univ.py", line 254, in prettyIn
    return int(value)
TypeError: int() argument must be a string, a bytes-like object or a number, not 'NoneType'
```

Digging into the source, it looks like `DnAttributes` can only be `True` or `False`, yet at line 111 in `operation/search.py` it's initialised with `None`, but is never set to `False` if `:dn` isn't found in the filter string.

To try and get around this, I tried adding `:dn` to the search string (so `memberOf:dn:1.2.840...`), which now doesn't throw an exception but still doesn't work.

For what it's worth, running the same query using `ldapsearch` from the commandline returns the results I'd expect (both with and without the `:dn` addition).

If there's any other information I can supply, please let me know!

Andy.

**Comment by root-locus** [comment]
I think I've faced with the same issue. At least, with the same error and same complex filter.

I've got a piece of code, where I'm trying to acquire all active users in AD with filter:
`FilterActiveUsers = "(&(objectCategory=person)(objectClass=user)(!(userAccountControl:1.2.840.113556.1.4.803:=2)))"`

This filter works perfectly with `ldapsearch` and with old version of pyasn1.

So, when I downgraded:
pyasn1: 0.4.2 -> 0.3.3
It works as intended. (upgrading/downgrading ldap3 2.3 <-> 2.4 didn't change anything)

I'm feared to dig all these search.py/univ.py/etc trying to understand how to format a proper asn.1 string with extensibleMatch from RFC2251. But this could be used as a temporary workaround or a hint where to start digging.

BTW, @ve7cxz could you please check the trick with versions? I think, LDAP_MATCHING_RULE_IN_CHAIN should be proceed the same as LDAP_MATCHING_RULE_BIT_AND

**Comment by etingof** [comment]
I suspect this issue is caused by removal of  `None` initializer special treatment in pyasn1 types constructors. Before 0.4.x you could do:

    DnAttributes(None)

Starting pyasn1 0.4.x you can only do:

    DnAttributes()
    # ... which is equivalent to
    DnAttributes(pyasn1.type.univ.noValue)

Since `DnAttributes` type has a default value (`False`), it will be used as initializer in all the cases above. So the fix would be either:

    if filter_node.assertion['dnAttributes'] is not None:
        matching_filter['dnAttributes'] = DnAttributes(filter_node.assertion['dnAttributes'])
    else:
        matching_filter['dnAttributes'] = DnAttributes()

Or better initilaize `filter_node.assertion['dnAttributes']` to `pyasn1.type.univ.noValue` instead of `None` by default. Would it be possible with ldap3?

**Comment by cannatag** [comment]
Hi,
thanks for reporting this issue. I'll try to follow the advice of Ilya in the next version of ldap3.

**Comment by fishhead108** [comment]
The same issue.

I've noticed that pyasn1: 0.4.2 on Python 2.7.12, 2.7.14 and 3.6.3, works well, but crash on 3.5.2 and 3.5.4.

**Comment by cannatag** [comment]
Please check code in dev. It should work again.

**Comment by gvialetto** [comment]
Hi,
I've checked it out with a fresh clone with this query:
```
(&(objectClass=group)(member:1.2.840.113556.1.4.1941:=<user DN>))
```
And I can confirm it works in the `dev` branch and returns the expected results.

**Comment by cannatag** [comment]
Thanks, will release the fix in the next version.

**Comment by cannatag** [comment]
Released in version 2.4.1
